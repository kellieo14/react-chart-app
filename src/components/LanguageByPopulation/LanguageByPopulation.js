import React, {useState, useEffect} from 'react';
import VerticalBar from '../Graphs/VerticalBar/VerticalBar';
import {sort} from '../../Utils';

export  const extractLanguages = (countries, languages) => {
    for (let country in countries) {
        countries[country].languages.forEach(element => {
            languages.push([element.name, Math.round(countries[country].population / 1000000)]);
        })
    }
}

export const sortLanguagesByName = (languages) => {
    languages.sort((a, b) => {
        return a[0] > b[0] ? 1 : -1
    }) 
}

export const getLanguageCounts = (languages, result) => {
    for (let language in languages) {
        if (result[languages[language][0]] === undefined) {
            result[languages[language][0]] = languages[language][1];
        } else {
            result[languages[language][0]] += languages[language][1];
        }
    }
}

export const sortLanguagesByPopulation = (result, languageCountArr, labels, countryCount) => {
    for (let language in result) {
        languageCountArr.push([language, result[language]]);
    }
    sort(languageCountArr, 'lanugage', 8, labels, countryCount);

}

const setTopLanguagesByCountry = (labels, setData, countryCount) => {
    if (labels.length > 0) {
        setData({
            title: 'Most Common Languages by People Count',
            labels: labels,
            datasets: [
            {
                label: 'Number of People / Million',
                backgroundColor: 'rgb(141,198,63, .2)',
                borderColor: 'rgb(141,198,63, .6)',
                borderWidth: 1,
                hoverBackgroundColor: 'rgb(141,198,63, .6)',
                data: countryCount
            }
            ]
        })
    } 
}

const LanguageByPopulation = (props) => {
    const countries = props.countries;
    const [data, setData] = useState({});
    
    useEffect(() => {
        let languages = [];
        let labels = [];
        let countryCount = [];
        let result=[];
        let languageCountArr = [];

        extractLanguages(countries, languages);
        sortLanguagesByName(languages);
        getLanguageCounts(languages, result);
        sortLanguagesByPopulation(result, languageCountArr, labels, countryCount);
        setTopLanguagesByCountry(labels, setData, countryCount);
    }, [countries]);

    return (
        <div>
            <VerticalBar data={data}/>
        </div>
      );
}

export default LanguageByPopulation;