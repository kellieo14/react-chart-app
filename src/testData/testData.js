export const countries = [
    {
        name: 'China', 
        region: 'Asia', 
        languages: [
            { iso639_1: "zh", iso639_2: "zho", name: "Chinese", nativeName: "中文 (Zhōngwén)"}
        ], 
        population: 1377422166,
        area: 9640011,
    }, 
    {
        name: 'India', 
        region: 'Asia',
        languages: [
            {iso639_1: "hi", iso639_2: "hin", name: "Hindi", nativeName: "हिन्दी"},
            {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
        ], 
        population: 1295210000,
        area: 3287590,
    }, 
    {
        name: 'United States of America', 
        region: 'Americas', 
        languages: [
            {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
        ], 
        population: 323947000,
        area: 9629091,
    },
    {
        name: 'Nigeria', 
        region: 'Africa', 
        languages: [
            {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
        ], 
        population: 186988000,
        area: 923768,
    }, 
    {
        name: 'France', 
        region: 'Europe', 
        languages: [
            {iso639_1: "fr", iso639_2: "fra", name: "French", nativeName: "français"}
        ], 
        population: 66710000,
        area: 640679,
    }, 
    {
        name: 'Australia', 
        region: 'Oceania', 
        languages: [
            {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
        ], 
        population: 24117360,
        area: 7692024,
    }, {
        name: 'Polarland', 
        region: 'Polar', 
        languages: [
            {iso639_1: "es", iso639_2: "spa", name: "Spanish", nativeName: "Español"}
        ], 
        population: 11239004,
        area: 1284000,
    }, {
        name: 'Otherland', 
        region: 'Other', 
        languages: [
            {iso639_1: "es", iso639_2: "spa", name: "Spanish", nativeName: "Español"}
        ], 
        population: 100,
        area: 619745,
    }
];