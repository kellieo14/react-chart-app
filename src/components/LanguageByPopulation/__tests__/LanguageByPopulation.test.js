import React from 'react';
import * as reactHooks from 'react';
import Enzyme, {mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import TestRenderer from 'react-test-renderer';
import { countries } from '../../../testData/testData';
import LanguageByPopulation, { extractLanguages, sortLanguagesByName, getLanguageCounts, sortLanguagesByPopulation } from '../LanguageByPopulation';
Enzyme.configure({adapter: new Adapter()});

const utils = require('../../../Utils');
utils.sort = jest.fn();

describe('LanguageByPopulation', () => {
    let spyUseState;
    let spyUseEffect;
    let setMockData;

    beforeEach(() => {
        setMockData = jest.fn();
        spyUseState = jest.spyOn(reactHooks, 'useState')
            .mockImplementationOnce(() => [{}, setMockData]);
        spyUseEffect = jest.spyOn(reactHooks, 'useEffect');
    });

    const expectedLanguages = [
        ['Chinese', 1377], 
        ['Hindi', 1295], 
        ['English', 1295], 
        ['English', 324], 
        ['English', 187], 
        ['French', 67], 
        ['English', 24], 
        ['Spanish', 11], 
        ['Spanish', 0],
    ];

    const expectedSortedLanguages = [
        ['Chinese', 1377422166], 
        ['English', 24117360], 
        ['English', 186988000], 
        ['English', 323947000], 
        ['English', 1295210000], 
        ['French', 66710000], 
        ['Hindi', 1295210000], 
        ['Spanish', 100],
        ['Spanish', 11239004], 
    ];

    const expectedResult = {
        Chinese: 1377422166,
        English: 1830262360,
        French: 66710000,
        Hindi: 1295210000,
        Spanish: 11239104,
        
    }

    it('should match snapshot', () => {
        const testRenderer = TestRenderer.create(<LanguageByPopulation />);
        expect(testRenderer.toJSON()).toMatchSnapshot();
    });

    it('should call useState and useEffect', () => {
        mount(<LanguageByPopulation countries={countries} />);
        expect(spyUseState).toHaveBeenCalled();
        expect(spyUseEffect).toHaveBeenCalled();
    });

    it('should extract languages and population from countries', () => {
        let languages = [];
        extractLanguages(countries, languages);
        expect(languages).toEqual(expectedLanguages);
    });

    it('should sort languages alphabetically', () => {
        const languagesToSort = [
            ['Chinese', 1377422166], 
            ['Hindi', 1295210000], 
            ['English', 1295210000], 
            ['English', 323947000], 
            ['English', 186988000], 
            ['French', 66710000], 
            ['English', 24117360], 
            ['Spanish', 11239004], 
            ['Spanish', 100],
        ];
        sortLanguagesByName(languagesToSort);
        expect(languagesToSort).toEqual(expectedSortedLanguages);
    });

    it('should calculate total number of people that speak each language', () => {
        let result = {};
        getLanguageCounts(expectedSortedLanguages, result);
        expect(result).toEqual(expectedResult);
    });

    it('should call the child function', () => {
        let languageCountArr = [];
        let labels = [];
        let countryCount = [];
        sortLanguagesByPopulation(expectedResult, languageCountArr, labels, countryCount);
        expect(utils.sort).toHaveBeenCalledTimes(2);
    });

    it('should sort languages by highest population', () => {
        const expectedLanguageCountArr = [
            ['Chinese', 1377422166], 
            ['English', 1830262360], 
            ['French', 66710000], 
            ['Hindi', 1295210000], 
            ['Spanish', 11239104]
        ];
        let languageCountArr = [];
        let labels = [];
        let countryCount = [];
        sortLanguagesByPopulation(expectedResult, languageCountArr, labels, countryCount);
        expect(languageCountArr).toEqual(expectedLanguageCountArr);

    });
})