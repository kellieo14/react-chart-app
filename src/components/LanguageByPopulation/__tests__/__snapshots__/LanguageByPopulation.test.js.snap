// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageByPopulation should match snapshot 1`] = `
<div>
  <div
    className="verticalBar"
  >
    <h4
      className="chartTitle"
    />
  </div>
</div>
`;
